### PR TITLE
ENH: Make Incremental work in grid search

### PR DESCRIPTION
**Pull request by TomAugspurger**
Closes https://github.com/dask/dask-ml/issues/195

Summary of changes

1. ``**kwargs`` in the Incremental constructor now refers to hyperparameters
   of `estimator`, rather than fit_kwargs
2. fit_kwargs are now passed in `fit`.

This *seemed* to work just fine with `sklearn.model_selection.GridSearchCV`.
Incremental.fit is correctly getting Dask (not NumPy) arrays.
Will try it out on a larger test now.

### CODE REVIEW

File: dask_ml/wrappers.py [hunk 1]
```diff
@@ -256,17 +259,16 @@ class Incremental(ParallelPostFit):
     >>> import sklearn.linear_model
     >>> X, y = make_classification(chunks=25)
     >>> est = sklearn.linear_model.SGDClassifier()
-    >>> clf = Incremental(est, classes=[0, 1])
-    >>> clf.fit(X, y)
+    >>> clf = Incremental(est)
+    >>> clf.fit(X, y, classes=[0, 1])
     """
     def __init__(self, estimator, **kwargs):
+        estimator.set_params(**kwargs)
```
Comment (TomAugspurger) [inline dask_ml/wrappers.py]: I think that calling `estimator.set_params` is the least surprising thing here. In general, I don't expect users to be passing additional kwargs here, just scikit-learn.

File: dask_ml/wrappers.py [hunk 2]
```diff
@@ -275,20 +277,38 @@ def fit(self, X, y=None):
...
+    def __repr__(self):
+        # Have to override, else all the parameters of estimator
+        # are duplicated
+        estimator = repr(self.estimator)
+        class_name = self.__class__.__name__
+        return '{}({})'.format(class_name, estimator)
+
+    def get_params(self, deep=True):
+        out = self.estimator.get_params(deep=deep)
+        out['estimator'] = self.estimator
```
Comment (TomAugspurger) [inline dask_ml/wrappers.py]: Awkward API: what if the estimator parameter has an `estimator` parameter? Raise an error?

**Comment by TomAugspurger** [comment]
This isn't quite working with the distributed scheduler yet.

**Comment by TomAugspurger** [comment]
This grew quite a bit, so here's a summary

1. ParellelPostFit.score & Incremental.score now use dask-aware scorers. Previously they materialized a single ndarray. This exposed a difficulty with the scikit-learn API that I haven't resolved yet: given an estimator like `sklearn.linear_model.SGDClassifier`, there's no way to inspect what underlying metric `SGDClassfier` actually uses. I happen to know that it uses `accuracy_score`, but I would like to detect that in code, so that we can swap out the scorer with a dask-aware one. Writing metrics that work on either NumPy or dask code would also work (and may be doable.) I'll think on it more and raise a scikit-learn issue. For now we ask the users to provide a `scoring` parameter to `Incremental` and `ParallelPostFit`.
2. I'm seeings some strange errors with the distributed scheduler, that I haven't narrowed down yet. When running `GridSearchCV(Incremental(SGDClassifier))` I get exceptions in the scheduler like

```pytb
distributed.client - WARNING - Couldn't gather 2 keys, rescheduling {"('getitem-bd5f254a6b4dabd84f4bf11f7b61edd7', 1)": [], "('getitem-bd5f254a6b4dabd84f4bf11f7b61edd7', 0)": []}
distributed.core - ERROR - "('astype-lt-getitem-4f528eaacb9ddb9447dec5cd09677027', 1)"
Traceback (most recent call last):
  File "/Users/taugspurger/sandbox/distributed/distributed/core.py", line 375, in handle_stream
    handler(**merge(extra, msg))
  File "/Users/taugspurger/sandbox/distributed/distributed/scheduler.py", line 1355, in update_graph
    ts = self.tasks[key]
KeyError: "('astype-lt-getitem-4f528eaacb9ddb9447dec5cd09677027', 1)"
distributed.core - ERROR - "('astype-lt-getitem-4f528eaacb9ddb9447dec5cd09677027', 1)"
Traceback (most recent call last):
  File "/Users/taugspurger/sandbox/distributed/distributed/core.py", line 321, in handle_comm
    result = yield result
  File "/Users/taugspurger/Envs/dask-dev/lib/python3.6/site-packages/tornado/gen.py", line 1099, in run
    value = future.result()
  File "/Users/taugspurger/Envs/dask-dev/lib/python3.6/site-packages/tornado/gen.py", line 1113, in run
    yielded = self.gen.send(value)
  File "/Users/taugspurger/sandbox/distributed/distributed/scheduler.py", line 1922, in add_client
    yield self.handle_stream(comm=comm, extra={'client': client})
  File "/Users/taugspurger/Envs/dask-dev/lib/python3.6/site-packages/tornado/gen.py", line 1099, in run
    value = future.result()
  File "/Users/taugspurger/Envs/dask-dev/lib/python3.6/site-packages/tornado/gen.py", line 1113, in run
    yielded = self.gen.send(value)
  File "/Users/taugspurger/sandbox/distributed/distributed/core.py", line 375, in handle_stream
    handler(**merge(extra, msg))
  File "/Users/taugspurger/sandbox/distributed/distributed/scheduler.py", line 1355, in update_graph
    ts = self.tasks[key]
KeyError: "('astype-lt-getitem-4f528eaacb9ddb9447dec5cd09677027', 1)"
Traceback (most recent call last):
  File "/Users/taugspurger/sandbox/distributed/distributed/client.py", line 1429, in _gather
    st = self.futures[key]
KeyError: "('getitem-4f528eaacb9ddb9447dec5cd09677027', 2)"
```

https://gist.github.com/f65705bd1605bee283741c18eef34b10 reliably has failures, but I haven't worked to minimize the example yet.

@mrocklin if I'm still struggling with the distributed issues in an hour or so, could I bother you for a debugging session?

**Comment by TomAugspurger** [comment]
Here's the GridSearchCV(Incremental(SGDClassifier)) on an 8GB dataset stored in Zarr, run locally.

https://gist.github.com/8ef23ae0861037637c03bbdaf56d0260

The GridSearch `n_jobs` is set to 1. We still get some parallelism, I think in loading data and scoring. Each call to the underlying `.partial_fit` is sequential still, since `n_jobs=1`.

https://streamable.com/l6k82

**Comment by TomAugspurger** [comment]
@jakirkham Once I get the GridSearchCV working with the distributed joblib backend, I'm going to turn the to the neuroimager data.

Do you have any simple scikit-image operations to apply to the dataset that would reduce some of the noise we saw? This would make for a nice recap blogpost from the sprint.

**Comment by TomAugspurger** [comment]
There was a bug in the previous implementation, as the scorer was reverting back to the underlying estimators score (`scoring` wasn't being passed through in get / set params).

It's fixed things so that https://gist.github.com/8ef23ae0861037637c03bbdaf56d0260 run successfully, though the larger zarr-based example still fails.

**Comment by mrocklin** [comment]
I'd be happy to take a look at this, but will be unfortunately inaccessible
for most of today.  I may have an hour or two in the late afternoon.  If
this is the case then I'll reach out to you off-list @TomAugspurger.
Alternatively I have much more time tomorrow.

**Comment by TomAugspurger** [comment]
No rush at all.

Given that this works fine for datasets that are persisted in (maybe distributed) memory, I'm going to merge this and followup on the distributed issues later.

**Pull request merged**